iothub/device/src/Transport/Mqtt/MqttTransportHandler.cs

```diff
@@ -4,1310 +4,1097 @@
+
+            if (_mqttTransportSettings.HasWill && _mqttTransportSettings.WillMessage != null)
             {
-                _channelFactory = settings.Protocol switch
+                _mqttClientOptionsBuilder.WithWillTopic(_deviceToCloudMessagesTopic);
+                _mqttClientOptionsBuilder.WithWillPayload(_mqttTransportSettings.WillMessage.Payload);
+
+                if (_mqttTransportSettings.WillMessage.QualityOfService == QualityOfService.AtMostOnce)
                 {
-                    IotHubClientTransportProtocol.Tcp => CreateChannelFactory(connectionCredentials, _additionalClientInformation, settings),
-                    IotHubClientTransportProtocol.WebSocket => CreateWebSocketChannelFactory(connectionCredentials, _additionalClientInformation, settings),
-                    _ => throw new InvalidOperationException($"Unsupported transport setting: {settings.Protocol}"),
-                };
+                    _mqttClientOptionsBuilder.WithWillQualityOfServiceLevel(MqttQualityOfServiceLevel.AtMostOnce);
+                }
+                else if (_mqttTransportSettings.WillMessage.QualityOfService == QualityOfService.AtLeastOnce)
+                {
+                    _mqttClientOptionsBuilder.WithWillQualityOfServiceLevel(MqttQualityOfServiceLevel.AtLeastOnce);
+                }
             }
 
-            _closeRetryPolicy = new RetryPolicy(new TransientErrorIgnoreStrategy(), 5, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
+            _mqttClient.ApplicationMessageReceivedAsync += HandleReceivedMessage;
+            _mqttClient.DisconnectedAsync += HandleDisconnection;
+
+            publishingQualityOfService = _mqttTransportSettings.PublishToServerQoS == QualityOfService.AtLeastOnce
+                ? MqttQualityOfServiceLevel.AtLeastOnce : MqttQualityOfServiceLevel.AtMostOnce;
+
+            receivingQualityOfService = _mqttTransportSettings.ReceivingQoS == QualityOfService.AtLeastOnce
+                ? MqttQualityOfServiceLevel.AtLeastOnce : MqttQualityOfServiceLevel.AtMostOnce;
         }
 
-        public TransportState State => (TransportState)Volatile.Read(ref _state);
-        public override bool IsUsable => State != TransportState.Closed && State != TransportState.Error;
-        public TimeSpan TwinTimeout { get; set; } = s_defaultTwinTimeout;
+        private bool certificateValidationHandler(MqttClientCertificateValidationEventArgs args)
+        {
+            return _mqttTransportSettings.RemoteCertificateValidationCallback.Invoke(
+                _mqttClient,
+                args.Certificate,
+                args.Chain,
+                args.SslPolicyErrors);
+        }
 
         #region Client operations
 
         public override async Task OpenAsync(CancellationToken cancellationToken)
         {
-            try
+            string clientId = _moduleId == null ? _deviceId : $"{_deviceId}/{_moduleId}";
```

Comment:
Here you only check for null but on 277 you check for whitespace too. We should be consistent.

Perhaps we should throw an invalid operation exception if it is whitespace?